### PR TITLE
Forbid the use of `#[target_feature]` on `main`

### LINKED ISSUE #108645: `#[target_feature]` is allowed on `main`
```rust
#[target_feature(enable = "avx2")] // no error
fn main() {}
```

This is a consequence of `#[target_feature]` being allowed on safe functions with `#![feature(target_feature_11)]`.

cc https://github.com/rust-lang/rust/issues/69098
@rustbot label T-lang T-compiler C-bug I-unsound F-target_feature_11

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Fixes #108645.

### CODE REVIEW

**Review by Noratrieb — APPROVED** [review]
r? Nilstrieb
@bors delegate
r=me when CI is green

**Comment by WaffleLapkin** [comment]
_waffle rust log analyzer_

There are two failures:
1. [`tests/ui/asm/x86_64/issue-89875.rs`](https://github.com/rust-lang/rust/blob/18caf88956ecf454e24307e598b8ac9967f10b07/tests/ui/asm/x86_64/issue-89875.rs) uses `#[target_feature]` on `main` (moving `main`'s inside and `#[target_feature]` to a separate fn should fix the test)
2. [`tests/ui/entry-point/imported_main_from_extern_crate.rs`](https://github.com/rust-lang/rust/blob/18caf88956ecf454e24307e598b8ac9967f10b07/tests/ui/entry-point/imported_main_from_extern_crate.rs) ICEs because the `main` function there is imported from another crate and the `#[target_feature]` attribute is local-only (probably an error?...)

**Comment by WaffleLapkin** [comment]
Hm, `target_feature` was made `local_only` in #98450 as an optimization...

Ig it is fine, actually? Maybe? If the codegen only happens inside the crate, then `target_feature` being local is totally fine. At least we don't have any tests that tried to access `target_feature` in non-local crate...

**Comment by Nilstrieb** [comment]
[This isn't unsound](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=86d1e418db983ce1cd8a2fd2c2ebe2eb), the safeaty check probably goes through target feature queries instead of the attr.
So I guess you should try figuring out how that works and whether you can use the same way - although encoding target_feature is also a possible solution.

**Comment by LeSeulArtichaut** [comment]
> the safeaty check probably goes through target feature queries instead of the attr

Yes, see https://github.com/rust-lang/rust/blob/18caf88956ecf454e24307e598b8ac9967f10b07/compiler/rustc_mir_transform/src/check_unsafety.rs#L363-L365

**Comment by LeSeulArtichaut** [comment]
Also we need to forbid `#[target_feature]` on `start`, no?

**Comment by Nilstrieb** [comment]
Now it looks good and green 😊
@bors r+

**Comment by matthiaskrgr** [comment]
failed in a rollup

https://github.com/rust-lang-ci/rust/actions/runs/4316566308/jobs/7532513148#step:26:17181

**Comment by compiler-errors** [comment]
@bors r-

Fair chance this will land after #108654, so this needs blessing and fixing.

**Comment by LeSeulArtichaut** [comment]
Rebased and fixed tests

**Comment by matthiaskrgr** [comment]
@bors r- failed in a rollup

https://github.com/rust-lang/rust/pull/108701
https://github.com/rust-lang-ci/rust/actions/runs/4326386535/jobs/7553739879#step:26:17204

**Comment by Nilstrieb** [comment]
these tests really liked putting `#[target_feature]` on main, lol

**Comment by LeSeulArtichaut** [comment]
Since [`#[target_feature]` is allowed on safe functions](https://doc.rust-lang.org/reference/attributes/codegen.html#wasm32-or-wasm64) (on stable), I guess we should allow `#[target_feature]` on `main`/`start`? It's sound and it would otherwise be a breaking change.

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 1]
```diff
@@ -283,6 +283,14 @@ fn check_main_fn_ty(tcx: TyCtxt<'_>, main_def_id: DefId) {
         error = true;
     }
 
+    if !tcx.codegen_fn_attrs(main_def_id).target_features.is_empty()
+        && !tcx.sess.target.is_like_wasm
+        && !tcx.sess.opts.actually_rustdoc
+    {
```
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/lib.rs:289]: is there any way you can avoid this actually_rustdoc check? these should not be added 😅
maybe overwrite the codegen_fn_attrs query in rustdoc? Or why exactly was this needed in the first place?

Comment (LeSeulArtichaut) [reply compiler/rustc_hir_analysis/src/lib.rs:289]: I basically copied-pasted the condition that allows `#[target_feature]` on safe functions (see the rationale for `actually_rustdoc` at the bottom of the comment): https://github.com/rust-lang/rust/blob/104f4300cfddbd956e32820ef202a732f06ec848/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L227-L244

Comment (Nilstrieb) [reply compiler/rustc_hir_analysis/src/lib.rs:289]: I guess it's okay then for now, we just need to make sure that they are both fixed at the same time. Could you add a comment to the one you added?

**Comment by LeSeulArtichaut** [comment]
(Committed a submodule change by accident, sorry for the noise)

**Pull request merged**